pkg/plugins/manager/pipeline/stages/bootstrap/bootstrap.go

```diff
@@ -0,0 +1,72 @@
+package bootstrap
+
+import (
+	"context"
+
+	"github.com/grafana/grafana/pkg/plugins"
+	"github.com/grafana/grafana/pkg/plugins/config"
+	"github.com/grafana/grafana/pkg/plugins/log"
+	"github.com/grafana/grafana/pkg/plugins/manager/loader/assetpath"
+	"github.com/grafana/grafana/pkg/plugins/manager/signature"
+)
+
+type Bootstrapper interface {
+	Bootstrap(ctx context.Context, src plugins.PluginSource, bundles []*plugins.FoundBundle) ([]*plugins.Plugin, error)
+}
+
+type ConstructFunc func(ctx context.Context, src plugins.PluginSource, bundles []*plugins.FoundBundle) ([]*plugins.Plugin, error)
+
+type DecorateFunc func(ctx context.Context, p *plugins.Plugin) (*plugins.Plugin, error)
```

Comment:
here I would make each "step" to implement the same interface. For the bootstrap steps, it could be something like:

```go
type BootstrapStep func(
  ctx context.Context,
  src plugins.PluginSource,
  bundles []*plugins.FoundBundle,
  prev []*plugins.Plugin,
) ([]*plugins.Plugin, error)
```

That way you just iterate over the steps, feeding the result of the previous step to the next one. This, in theory, removes the dependency between steps, at least generally. The order should still matter obviously since some steps will require a previous step to be run first. Also, steps can ignore the inputs that they don't need.

The inputs should be well thought though, if we need to add an input in the future, we would need to change all the steps, which can be tedious.

Reply:
another possibility is to store the original inputs in the `Bootstrap` struct, so steps can access things like `bootstrap.Src` and not have it as part of the Step definition (so we can have a simpler Step function like `type Step func(ctx context.Context, prev []*plugins.Plugin) ([]*plugins.Plugin, error)`)

Reply:
Interesting idea. Will try explore with it. My initial reaction is I fear that it may be providing too much flexibility where someone might trip themselves up with it. But will check it out to really see

Reply:
>The order should still matter obviously since some steps will require a previous step to be run first.

I think this is part of the problem with the pattern for me. Having one signature for all steps can make it confusing what to expect across the different steps. The dependency to me is intrinsic to the pattern, where you should have to think a bit more carefully about the big picture. For example, having `prev []*plugins.Plugin` as an argument in the construct func doesn't make sense as there's no real plugin created yet. 

>The inputs should be well thought though, if we need to add an input in the future, we would need to change all the steps, which can be tedious.

Which is why I like the idea of having specific funcs doing specific things, and limits the blast radius if we need to change the API. I agree the order/dependency part can be confusing, which is why I was thinking that we could have a signature that takes steps in the order in which they are executed. Maybe subsequent stages will be more fitting to this pattern though? 

Aside from all that, I think that since each stage is an interface, people could just implement it however they want. It means they'll have more to write, rather than just configuring the steps, but at least that's an option. That way they can setup their struct however they want. It'd be great to avoid that obviously, but the option is there if needed. We can always adapt to our needs more easily in this case at least since this is an "internal" API - so I don't think we need to get it perfect to get something working.

Reply:
The problem I see with having one custom signature function per step is:

 - It's more complex to write a new step, you need to think about the function definition (e.g. `DecorateFuncs`) and modify the stage struct (e.g. `Bootstrap.decorateSteps`) rather than just re-apply the same pattern each time you want a new step.
 - It's more difficult to change. If now you want to move a step that was after `decorate`, and place it before, you will likely to change the functions (inputs/outputs) for the other steps. If all the steps apply the same interface, they are easily interchangeable. 

As long as we maintain the order of the step list, steps may be able to rely on other steps. We can obviously add validation to a step, and return an error if some information is missing. A step "B" may need the step "A" to be run first, but don't care if there are more or less steps between them. That's something nice you can rely on if steps are generic.

The first step obviously will be a special case but I think that's fine.

I am thinking on the middleware pattern for example. I think it can be nicely applied here.

Reply:
>It's more complex to write a new step, you need to think about the function definition (e.g. DecorateFuncs) and modify the stage struct (e.g. Bootstrap.decorateSteps) rather than just re-apply the same pattern each time you want a new step.

I think by having well defined steps, it makes it clearer how to build and compose the stage. Just having the same function signatures makes me wonder, why would I need to create more than one if they are all the same?

>As long as we maintain the order of the step list, steps may be able to rely on other steps.

I think this is a key part of what I have tried to do, but maybe your approach does sound more like middlewares. I would favour being more explicit in what you have to build rather than generic. I think flexibility is great, but at the end of the day it gets harder and harder to reason about. Do you want to try demonstrate what this might look like if you branch off from my branch if you think they are closely related? Or maybe just start another to demonstrate the idea. It might be easier to compare that way I think!

Reply:
that makes sense :+1: will try to allocate some time for this to try it out how it would look like

Reply:
proposal here! https://github.com/grafana/grafana/pull/72028

basically what I like is that you can apply the same mental pattern for each stage / step

the problem is that the stage struct becomes a blob of things different from each other (even though grouped) but I still prefer to handle custom inputs in the struct vs in the different step function definiton.